src/v/storage/log_reader.cc

```diff
@@ -430,7 +433,20 @@ log_reader::do_load_slice(model::timeout_clock::time_point timeout) {
               return ss::make_ready_future<storage_t>(
                 std::move(batches_filled));
           }
+          // To keep things consistent, our internal accounting is all done in
+          // untranslated offsets, even if we've been requested to return
+          // translated offsets.
           _expected_next = model::next_offset(batches.back().last_offset());
+
+          if (_config.translate_offsets) {
+              vassert(
+                _translator, "Expected offset translactor to be initialized");
+              for (auto& b : batches) {
+                  b.header().base_offset = _translator->from_log_offset(
+                    b.base_offset());
+              }
+          }
```

Comment:
nice. i wonder if at some point this could become cheap enough that'd we could always translate the offset into a new b.translated_offset()

Reply:
Yeah in the future once getting the translated offset is a matter of reading deserializing the metadata from an entry, it seems plausible that that offset gets put into the `ctx` as well, along with the term.